### PR TITLE
Allow to use a custom cache instead of using the filesystem only

### PR DESCRIPTION
**Pull request by Paulomart**
This commit adds the ability to specify a cache factory function instead of a folder
name only when creating an Authflow instance.

The cache factory function will take additional options and must return a cache. A cache
is an class instance that supports basic read and write operations.

The old behavior of the cache has been implemented by a FileCache. The FileCache will be
used as default and works like the old implementation used to.

### CODE REVIEW

**Comment by rom1504** [comment]
The code looks reasonable.
There is no automated test in this repo so we have to test manually. Can you try to run what's in test folder and report if it works ?
Maybe you could also add some more manual tests in that folder that could test if the code you added works as you expect

File: src/TokenManagers/MsaTokenManager.js [hunk 1]
```diff
@@ -1,23 +1,21 @@
...
 class MsaTokenManager {
-  constructor (msalConfig, scopes, cacheLocation) {
+  constructor (msalConfig, scopes, cache) {
     this.msaClientId = msalConfig.auth.clientId
     this.scopes = scopes
-    this.cacheLocation = cacheLocation || path.join(__dirname, './msa-cache.json')
+    this.cache = cache
 
-    this.reloadCache()
+    // this.reloadCache()
```
Comment (Kashalls) [inline src/TokenManagers/MsaTokenManager.js:10]: If we removed this function, you can remove this.

File: src/common/cache/FileCache.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+  async setCachedPartial (cached) {
+    await this.setCached({
+      ...this.cache,
+      ...cached
+    })
+  }
+}
+
+module.exports = { FileCache }
```
Comment (Kashalls) [inline src/common/cache/FileCache.js:39]: Should be `module.exports = FileCache` because you are not exporting more than one item.

**Review by Kashalls — APPROVED** [review]
Looks good to me, just a couple nick picks on my end.

**Comment by Kashalls** [comment]
> @rom1504 I executed the tests using `npm test`. Also the tests get executed by the CI: https://github.com/PrismarineJS/prismarine-auth/runs/4278337739?check_suite_focus=true#step:5:14
> 
> Additionally I integrated these changes into my application an also verified that this code works:
> 
> ```ts
> const flow = new Authflow(undefined, './test', { authTitle: Titles.MinecraftNintendoSwitch, deviceType: 'Nintendo' }, codeCallback);
> const response = await flow.getMinecraftJavaToken({ fetchEntitlements: true, fetchProfile: true });
> console.log(response);
> ```

Could you please write a test in the tests folder that covers the Cache system?

**Comment by Paulomart** [comment]
> Could you please write a test in the tests folder that covers the Cache system?

@Kashalls what kind of test are you thinking of? Just testing the FileTest, to ensure that the managers call the cache or something entirely different?

**Comment by Kashalls** [comment]
> > Could you please write a test in the tests folder that covers the Cache system?
> 
> @Kashalls what kind of test are you thinking of? Just testing the FileCache, ensure that the managers call the cache or something entirely different?

I was thinking of writing data to the cache and ensuring that it came out correctly. Obviously use false data :)

**Review by extremeheat — APPROVED** [review]

**Comment by Paulomart** [comment]
@Kashalls I added a test case for that. Please check if that fits.

**Comment by Kashalls** [comment]
> @Kashalls I added a test case for that. Please check if that fits.

Looks Fine To Me.

@extremeheat ?

**Comment by extremeheat** [comment]
Did you test it @Kashalls @LucienHH ?

**Comment by LucienHH** [comment]
> Did you test it @Kashalls @LucienHH ?

Yes, tested with a Redis solution in multiple scenarios and worked great, I did notice a bug when using sisuAuth however where when verifying the tokens it always returns false. This is because sisuAuth doesn't save the userToken into the cache, anyway, that's not an issue with this pr just something I need to fix later on.

**Comment by extremeheat** [comment]
Thanks

**Comment by Kashalls** [comment]
> > Did you test it @Kashalls @LucienHH ?
> 
> Yes, tested with a Redis solution in multiple scenarios and worked great, I did notice a bug when using sisuAuth however where when verifying the tokens it always returns false. This is because sisuAuth doesn't save the userToken into the cache, anyway, that's not an issue with this pr just something I need to fix later on.

Do you have an example of using it with Redis?

**Comment by LucienHH** [comment]
@kashalls

```js
const { Authflow, Titles } = require('prismarine-auth')
const Redis = require('ioredis');
const JSONRedis = require('redis-json');

class InMemoryCache {
    constructor(username, cacheName) {
        this.username = username,
        this.cacheName = cacheName
        this.uri = process.env.REDIS_URI.split(':');
        this.client = new Redis({
            port: parseInt(this.uri[1]),
            host: this.uri[0],
            password: process.env.REDIS_PASSWORD,
        });
        this.jsonClient = new JSONRedis(this.client);
    }
    async getCached() {
        return await this.jsonClient.get(`tokens:${this.username}:${this.cacheName}`) || {}
    }
    async setCached(value) {
        return await this.jsonClient.set(`tokens:${this.username}:${this.cacheName}`, value)
    }
    async setCachedPartial(value) {
        await this.jsonClient.set(`tokens:${this.username}:${this.cacheName}`, {
                ...await this.jsonClient.get(`tokens:${this.username}:${this.cacheName}`),
                ...value
        })

    }
}

const cacheFactory = ({ username, cacheName }) => new InMemoryCache(username, cacheName)

const doAuth = async () => {

    const flow = new Authflow('CacheTest', cacheFactory, { authTitle: Titles.MinecraftNintendoSwitch, Type: 'Nintendo' })

    await flow.getXboxToken()

    setInterval(async () => {

        const response = await flow.getXboxToken()

        console.log(response)

    }, 5000);
}

doAuth()
```

**Pull request merged**